Add PDO Metadata Storage Handler

tdiscuit opened the pull request:
This is to add support for storing of metadata in a SQL database.

The original code was taken from [issue #44](https://github.com/simplesamlphp/simplesamlphp/issues/44) and has been updated to work with the latest version of simpleSAMLphp and additional databases.

I have also included documentation and updated the config template to show usage.

jaimeperez commented:
Thanks for your contribution @tdiscuit!

I see you've done a pretty good and exhaustive job here :+1: 

I'll try to set this up asap to verify everything is in place, and take a quick look at the code before merging it. I'll let you know if I have any questions while reviewing.

tdiscuit commented:
Hey @jaimeperez 

I found a couple bugs that I overlooked during my initial pull request and made some updates and tested yesterday. Also wanted to check-in to see if you had any futher issues with this pull request.

jaimeperez commented:
No, unfortunately I haven't had time yet to take a look at this, sorry. I'll let you know as soon as I can get my hands on it.

jaimeperez commented on bin/importPdoMetadata.php:
```diff
@@ -0,0 +1,25 @@
+	if ("pdo" === $s['type']) {
+		$mdshp = new SimpleSAML_Metadata_MetaDataStorageHandlerPdo($s);
+		$mdshp->initDatabase();
+		foreach (glob("metadata/*.php") as $filename) {
```
I'm assuming this allows you to periodically import metadata into the DB, right? In that case, we might be assuming too much here if we expect all metadata to be in `metadata/{set}.php`, as that could change depending on the source. For instance, it could make sense to use this to pull metadata generated by metarefresh.

jaimeperez commented on bin/importPdoMetadata.php:
```diff
@@ -0,0 +1,25 @@
+# that a PDO source exists.
+foreach ($config['metadata.sources'] as $s) {
+	# If pdo is configured, create the new handler and add in the metadata sets.
+	if ("pdo" === $s['type']) {
```
This one would probably be more readable the other way around:

```php
if ($s['type'] === 'pdo']) {
```

jaimeperez commented on bin/initMDSPdo.php:
```diff
@@ -0,0 +1,27 @@
+		$metadataStorageHandler = new SimpleSAML_Metadata_MetaDataStorageHandlerPdo($source);
+		$result = $metadataStorageHandler->initDatabase();
+
+		if ($result === FALSE) {
```
We're also trying to get rid of all appearances of `TRUE`, `FALSE`, and `NULL` in uppercase.

jaimeperez commented on config-templates/config.php:
```diff
@@ -197,6 +197,61 @@
+     * Ensure that you have the required PDO database drive installed
+     * for your connection string.
+     */
+    'database.dsn' => 'mysql:host=localhost;dbname=saml',
```
Should these options be commented out? In the end, this is optional, so those copying and pasting the template should not have database options configured (not to talk about defaults) if they are not going to use a database, IMHO.

jaimeperez commented on docs/simplesamlphp-metadata-pdostoragehandler.txt:
```diff
@@ -0,0 +1,79 @@
+Introduction
+------------
+
+If you want to run a clustered simpleSAMLphp IdP service and you would like to have centralized storage for metadata, you can use the PDO metadata storage handler.
```
We're also trying to change this. In the old days we had a mixture with the caps in the name, but now we try to use always the capitalized version: **SimpleSAMLphp**.

jaimeperez commented on docs/simplesamlphp-metadata-pdostoragehandler.txt:
```diff
@@ -0,0 +1,79 @@
+Preparations
+------------
+
+You will need to have the appropriate PDO drivers for your database and you will have to configure the database secion within the config/config.php file.
```
s/secion/section/

jaimeperez commented on config-templates/config.php:
```diff
@@ -197,6 +197,61 @@
+     * Database
+     *
+     * This database configuration is optional. If you are not using
+     * core funcationlity or modules that require a database, you can
```
s/funcationlity/functionality/

jaimeperez commented on config-templates/config.php:
```diff
@@ -197,6 +197,61 @@
+
+    /*
+     * Database connection string.
+     * Ensure that you have the required PDO database drive installed
```
s/drive/driver/

jaimeperez commented on config-templates/config.php:
```diff
@@ -197,6 +197,61 @@
+    'database.prefix' => '',
+
+    /*
+     * True or false if you would like a persisent datata connection
```
s/persisent/persistent/

jaimeperez commented on config-templates/config.php:
```diff
@@ -197,6 +197,61 @@
+    /*
+     * Database slave configuration is optional as well. If you are only
+     * running a single database server, leave this blank. If you have
+     * a master/slave conifguration, you can define as many slave servers
```
s/conifguration/configuration/

jaimeperez commented on config-templates/config.php:
```diff
@@ -197,6 +197,61 @@
+     * a master/slave conifguration, you can define as many slave servers
+     * as you want here. Slaves will be picked at random to be queried from.
+     *
+     * Configuration options in the salve array are exactly the same as the
```
s/salve/slave/

jaimeperez commented on docs/simplesamlphp-metadata-pdostoragehandler.txt:
```diff
@@ -0,0 +1,79 @@
+
+	[root@simplesamlphp simplesamlphp]# php bin/initMDSPdo.php
+
+If you connect to your datbase, you will see 11 new empty tables; one for each metadata set.
```
s/datbase/database/

jaimeperez commented on docs/simplesamlphp-metadata-pdostoragehandler.txt:
```diff
@@ -0,0 +1,79 @@
+-------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ...
+https://openidp.feide.no | {"name":{"en":"Feide OpenIdP - guest users","no":"Feide Gjestebrukere"},"description":"Here you can login with your account on Feide RnD OpenID. If you do not already have an account on this identity provider, you can create a new one by following the create new account link and follow the instructions.","SingleSignOnService":"https:\/\/openidp.feide.no\/simplesaml\/sam ...
+
+There is an included script in the `bin` directory that will import all flatfile metadata files and store them in the database, but you can use an external tool to maintain the metadata in the database. This document will only cover adding metadata using the included utility, but the tables above should provide enough information if you would like to create a utility to manage your metadata exter ...
```
Ah, now I see I should have read the documentation first... :blush:

jaimeperez commented on lib/SimpleSAML/Database.php:
```diff
@@ -0,0 +1,248 @@
+ * @package simpleSAMLphp
+ */
+
+class SimpleSAML_Database {
```
I think we can use namespaces here. We've already done so for some of the new classes (or refactored ones), and this class is completely new, so I see no reason to stick to the old naming convention :smiley:

tdiscuit replied:
It definitely makes sense to use namespaces instead since that's the direction SSP is going. I'll update this to use a namespace.

jaimeperez commented on lib/SimpleSAML/Database.php:
```diff
@@ -0,0 +1,248 @@
+    /**
+	 * Retrieves the current database instance. Will create a new one if there isn't an existing connection.
+	 *
+	 * @param object $altConfig Optional: Instance of a SimpleSAML_Configruation class
```
s/Configruation/Configuration/

jaimeperez commented on lib/SimpleSAML/Database.php:
```diff
@@ -0,0 +1,248 @@
+	/**
+	 * Private constructor that restricts instantiation to getInstance().
+	 *
+	 * @param object $config Instance of the SimpleSAML_Configruation class
```
Copy/paste? :sweat_smile:

jaimeperez commented on lib/SimpleSAML/Metadata/MetaDataStorageHandlerPdo.php:
```diff
@@ -0,0 +1,226 @@
+
+
+	/**
+	 * This function loads the given set of metadata from a file our configured database.
```
*from a file* ***to*** *a configured database*?

jaimeperez commented on tests/SimpleSAML/DatabaseTest.php:
```diff
@@ -0,0 +1,245 @@
+	/**
+	 * @covers SimpleSAML_Database::read
+	 * @covers SimpleSAML_Database::query
+	 * @expectedException Exception
```
Would it make sense to have our own exception here? `DBException`?

tdiscuit replied:
I believe that would be a better option rather than using the generic Exception since most of SimpleSAMLphp uses exception handling. Do we want to include adding an extension of the built-in Exception class within this pull request, or handle that in a separate one which could include reworking `SimpleSAML_Error_Exception` to be a bit more flexible so we could have something along the lines of:

* \SimpleSAML\Exception\Database for database errors
* \SimpleSAML\Exception\Metadata for anything within the metadata storage sources/handlers
* \SimpleSAML\Exception\Session for session store issues
* \SimpleSAML\Exception\Generic
* etc.

At this point I'm inclined to say leave it as Exception for now and then create a separate pull request for exception handling as a whole, but I'm up for either or.

jaimeperez replied:
Hi!

> On 21 Jul 2015, at 18:40 pm, Tyler Antonio <notifications@github.com> wrote:
> I believe that would be a better option rather than using the generic Exception since most of SimpleSAMLphp uses exception handling. Do we want to include adding an extension of the built-in Exception class within this pull request, or handle that in a separate one which could include reworking SimpleSAML_Error_Exception to be a bit more flexible so we could have something along the lines of:
> 
> 	• \SimpleSAML\Exception\Database for database errors
> 	• \SimpleSAML\Exception\Metadata for anything within the metadata storage sources/handlers
> 	• \SimpleSAML\Exception\Session for session store issues
> 	• \SimpleSAML\Exception\Generic
> 	• etc.
> At this point I'm inclined to say leave it as Exception for now and then create a separate pull request for exception handling as a whole, but I'm up for either or.

I agree. After all, this PR is quite big already, so let’s leave it as an Exception now. We’ll have time to enhance the way we work with exceptions.

--
Jaime Pérez
UNINETT / Feide
mail: jaime.perez@uninett.no
xmpp: jaime@jabber.uninett.no

"Two roads diverged in a wood, and I, I took the one less traveled by, and that has made all the difference."
- Robert Frost

jaimeperez commented on tests/SimpleSAML/DatabaseTest.php:
```diff
@@ -0,0 +1,245 @@
+	 */
+	public function NoSuchTable()
+	{
+		$this->db->write("DROP TABLE phpunit_nonexistant", false);
```
s/nonexistant/nonexistent/

jaimeperez commented:
Impressive work Tyler! :smile: 

I've gone through it and looks pretty good. Didn't get deep into the code, but as you are providing tests, and they are passing seamlessly, I'm assuming it works like a charm.

So apart from a few things (you see most of my comments are just regarding typos), I think this is almost ready!

tdiscuit commented:
Thanks Jaime! :smile:

I've addressed some of your comments already, mainly the many spelling mistakes haha. I replied to a couple of your comments above and I'm working on the last couple things that you mentioned. Should be done addressing the other comments within a day or so.

tdiscuit commented:
Hey Jaime,

I've gone through and actioned all your comments and I just went through and updated the coding style to match the rest of SimpleSAMLphp where appropriate. I noticed there was a difference in style between the old class names and the new files that use namespaces so the SimpleSAML\Database class uses the new style while SimpleSAML_MetadataStorageHandlerPdo uses the old one.

I noticed coveralls is having an issue with the repo as it's not generating reports since one of my previous commits. Not sure if that's intended or not.. Travis is showing 100% again for the database once I fixed the namespace for PDOException and Exception so should be good there.

I believe we are good to go on this PR, let me know if I missed anything. I've been looking at the same code for a while now so there is a chance that I did something silly and keep overlooking it.

jaimeperez commented:
Merged! :smile: 

Not worried about the coveralls failures. They're having issues and went read-only, so that's why.

Thanks a lot again for all your work!

dredix84 commented:
Hello All,
My colleagues and I are keen on getting this working. I know it has been sometime since there was any activity in this area but wanted pickup where this was left off. @tdiscuit, what are the outstanding items/work to be done on this.
I have not tested as yet but it seem to be completed or close to completed.

jaimeperez commented:
Hi @dredix84!

This PR was merged, so the functionality has been in place for some time now, including the 1.14.x releases. However, I don't have experience myself using it, so maybe @tdiscuit can provide more information on that.

tdiscuit commented:
Hi @dredix84,

Is there something you had in mind that is missing or doesn't work? I've been using the PDO metadata storage service since it was merged on our production SSO environment. There is some documentation on using it [here](https://simplesamlphp.org/docs/stable/simplesamlphp-metadata-pdostoragehandler). Actually managing the metadata stored in the database is quite basic right now as it involves a script in the bin directory.

In my environment we've been using a modified version of that script that pulls metadata stored in individual files in a git repository, does some validation on the metadata sets so we have all the information that we want from a SP in there (ex. contacts, service name, signing/encryption certificates, etc.), and then adds to the database using the `addEntry` function in the `SimpleSAML_Metadata_MetaDataStorageHandlerPdo` class.

I was hoping one day to simplify this into a module that can be enabled and allow you to manage it through a web-ui, but I haven't really had the time to sit down and plan it out.

dredix84 commented:
Thanks for the quick reply.
There wasn't a feature I had in mind as I was wondering about the production readiness of the code.
Thanks for the link to the documentation as this will help immensely.

andre-dixon-qm commented:
Hello again all (same person as @dredix84 but now using work account). I have started to embark upon creating a class used for authsource management. I tested the existing code for database metadata management which I have successfully working  (with some minor issues). To make this less file dependent, we though it would be a good idea to have the authsource currently located in config/authsources.php be stored in the database for easier management.

A similar approach will be used to that of the metadata where there will be a key field and another field which the authsource will be stored in JSON format. I wanted to get the community input on this.

Current work (not yet submitted)
Class will be \lib\SimpleSAML\Authsource\AuthsourceHandler.php
The method used to get the auth source will be determined by a new entry in the config\config.php known as 'authsource.source'.
And example would be `'authsource.source' => 'file'` or `'authsource.source'  => 'pdo'`.
`'authsource.source' => 'file'` would tell the handler to retrieve data from the \lib\SimpleSAML\Authsource\Authsource_Sourcefile.php. In the case of pdo then Authsource_Sourcepdo.php would be loaded.

ee0pdt commented:
@andre-dixon-qm I'm very interested in your suggestion, did you ever follow up on that idea?

dredix84 commented:
Hello,
Sorry for the late reply. I know it has been a while and I have since left
that company which I was working on that project. Nonetheless, I did
complete the module which allowed simplesamplPhp to used alternate data for
the AuthSource.
It's something I can look back into independently if you are still
interested.

*André N. Dixon*

---
<http://www.dredix.net/>
Making technology work for you
*DreDix Inc* www.dredix.net

<http://www.wpspartans.com>
Your 24/7 WordPress support experts.
*WP Spartans* www.wpspartans.com


On Tue, Oct 24, 2017 at 3:56 AM, Pete Thorne <notifications@github.com>
wrote:

> @andre-dixon-qm <https://github.com/andre-dixon-qm> I'm very interested
> in your suggestion, did you ever follow up on that idea?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/simplesamlphp/simplesamlphp/pull/189#issuecomment-338906359>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACOf0hHwOd8EQ7yPX5gFM7Gxca2XGTsFks5svZg7gaJpZM4D8mVl>
> .
>

ee0pdt commented:
Hi André,

Thank you for getting back to me. If you have any code left over from that project that you're able to share I'd love to see it. We've built SimpleSAML into our SaaS product and it would be very helpful to have a way to add new configurations without having to touch code. Hence the PDO option would be very interesting to us.

regards

Pete

dredix84 commented:
Thanks.
I will look into this and pass code on.

On Nov 30, 2017 3:14 AM, "Pete Thorne" <notifications@github.com> wrote:

> Hi André,
>
> Thank you for getting back to me. If you have any code left over from that
> project that you're able to share I'd love to see it. We've built
> SimpleSAML into our SaaS product and it would be very helpful to have a way
> to add new configurations without having to touch code. Hence the PDO
> option would be very interesting to us.
>
> regards
>
> Pete
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/simplesamlphp/simplesamlphp/pull/189#issuecomment-348113908>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACOf0t0ZW4SnPbW_0zNxnZft5EjQpGGqks5s7mPYgaJpZM4D8mVl>
> .
>
